libyul/Utilities.cpp

```diff
@@ -86,56 +86,84 @@ std::string solidity::yul::reindent(std::string const& _code)
 	return out.str();
 }
 
-u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfNumberLiteral(std::string_view const _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::Number, "Expected number literal!");
-
-	static std::map<YulString, u256> numberCache;
-	static YulStringRepository::ResetCallback callback{[&] { numberCache.clear(); }};
+	return {LiteralValue::Data(_literal), std::string(_literal)};
+}
+LiteralValue solidity::yul::valueOfStringLiteral(std::string_view const _literal, bool const _unlimited)
+{
+	std::string const s(_literal);
+	return {u256(h256(s, h256::FromBinary, h256::AlignLeft)), s, _unlimited};
+}
+LiteralValue solidity::yul::valueOfBoolLiteral(std::string_view const _literal)
+{
+	if (_literal == "true")
+		return {1, "true"};
+	else if (_literal == "false")
+		return {0, "false"};
 
-	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
-	if (isNew)
+	yulAssert(false, "Unexpected bool literal value!");
+}
+LiteralValue solidity::yul::valueOfLiteral(std::string_view const _literal, LiteralKind const& _kind, bool const _unlimited)
+{
+	switch (_kind)
 	{
-		std::string const& literalString = _literal.value.str();
-		yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-		it->second = u256(literalString);
+	case LiteralKind::Number:
+		return valueOfNumberLiteral(_literal);
+	case LiteralKind::Boolean:
+		return valueOfBoolLiteral(_literal);
+	case LiteralKind::String:
+		return valueOfStringLiteral(_literal, _unlimited);
+	default:
+		yulAssert(false, "Unexpected literal kind!");
 	}
-	return it->second;
 }
 
-u256 solidity::yul::valueOfStringLiteral(Literal const& _literal)
+std::string solidity::yul::formatLiteral(solidity::yul::Literal const& _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::String, "Expected string literal!");
-	yulAssert(_literal.value.str().size() <= 32, "Literal string too long!");
+	std::string result;
+	// literal values are immutable, so if there's a hint, we better take it; invalid hints are taken care of by
+	// AsmAnalysis (string too long, number too large, ...)
+	if (_literal.value.hint())
+		result = *_literal.value.hint();
+	else
+		if (_literal.kind == LiteralKind::Boolean)
+		{
+			yulAssert(_literal.value() == 0 || _literal.value() == 1, "Could not format boolean literal");
+			result = _literal.value() == 1 ? "true" : "false";
+		}
+		else
+			result = _literal.value().str();
```

Comment:
This looks like you're formatting strings and integers the same way. How can that work? I.e. how does it preserve the distinction between `1234` and `"1234"`?

Actually, does the hint include the quotes? Because if it does not, then I can't see how we can distinguish values like `'1234'` and `"1234"` or `'1234'` and `hex'1234'`.

Also, if there's no distinction, the value cannot be unambiguously recovered from the hint (which matters if we want to enforce that they match).

Reply:
The quoting and escaping of string literals is done in `AsmPrinter`, so the `LiteralKind` is also always required to unambiguously print and/or recover a literal. In your example you'd get

```
object "A" {
    code {
        pop(datasize(hex'616263'))
    }
    data 'abc' "1234"
}
// ----
// step: disambiguator
//
// object "A" {
//     code { pop(datasize("abc")) }
//     data "abc" hex"31323334"
// }

```

Reply:
Oh, so we're actually not preserving the representation perfectly? That's surprising. I had the impression that the AST allowed us to mostly reproduce the original source aside from comments and whitespace and that the representation hint was meant to help with that, but apparently not. I see from your output that it doesn't even keep hex numbers as hex so it seems kinda useless.

I wonder what we'd really lose by removing it.

Reply:
Yeah, agree, this can probably be removed and hopefully make for a much cleaner data structure without losing too much